### PR TITLE
loggedInUser issue fixes #4372

### LINKED ISSUE #4372: RestApiUtil.getLoggedInUsername(); always returns admin

### PR DESCRIPTION
**Pull request by Krishanx92**
### Proposed changes in this pull request

- resolve #4372
- new "request" param introduce to RestApiUtil.java -"getLoggedInUsername" method.
- services and testcases change according to that.
- added new "loggedInUser" param to getETag().
- added getEndUserName() to OAuth2Authenticator.
- added new test case for OAuth2Authenticator.

### When should this PR be merged

[Please describe any preconditions that need to be addressed before we
can merge this pull request.]

### Follow up actions

[List any possible follow-up actions here; for instance, testing data
migrations, software that we need to install on staging and production
environments.]

-

#### General

- [ ] **Is this PR explained thoroughly?** All code changes must be accounted for in the PR description.
- [ ] **Is the PR labeled correctly?**

#### Functionality

- [ ] **Are all requirements met?** Compare implemented functionality with the requirements specification.
-[ ] **Does the UI work as expected?** There should be no Javascript errors in the console; all resources should load. There should be no unexpected errors. Deliberately try to break the feature to find out if there are corner cases that are not handled.

#### Code

- [ ] **Do you fully understand the introduced changes to the code?** If not ask for clarification, it might uncover ways to solve a problem in a more elegant and efficient way.
- [ ] **Does the PR introduce any inefficient database requests?** Use the debug server to check for duplicate requests.
- [ ] **Are all necessary strings marked for translation?** All strings that are exposed to users via the UI must be [marked for translation](https://docs.djangoproject.com/en/1.10/topics/i18n/translation/).

#### Tests

- [ ] **Are there sufficient test cases?** Ensure that all components are tested individually; models, forms, and serializers should be tested in isolation even if a test for a view covers these components.
- [ ] **If this is a bug fix, are tests for the issue in place?**  There must be a test case for the bug to ensure the issue won’t regress. Make sure that the tests break without the new code to fix the issue.
- [ ] **If this is a new feature or a significant change to an existing feature?** has the manual testing spreadsheet been updated with instructions for manual testing?

#### Security

- [ ] **Confirm this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets.**
- [ ] **Are all UI and API inputs run through forms or serializers?**
- [ ] **Are all external inputs validated and sanitized appropriately?**
- [ ] **Does all branching logic have a default case?**
- [ ] **Does this solution handle outliers and edge cases gracefully?**
- [ ] **Are all external communications secured and restricted to SSL?**

#### Documentation

- [ ] **Are changes to the UI documented in the platform docs?** If this PR introduces new platform site functionality or changes existing ones, the changes should be documented.
- [ ] **Are changes to the API documented in the API docs?** If this PR introduces new API functionality or changes existing ones, the changes must be documented.
- [ ] **Are reusable components documented?** If this PR introduces components that are relevant to other developers (for instance a mixin for a view or a generic form) they should be documented in the Wiki.

### CODE REVIEW

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.commons/src/main/java/org/wso2/carbon/apimgt/rest/api/common/impl/OAuth2Authenticator.java [hunk 1]
```diff
@@ -112,6 +115,18 @@ private boolean validateTokenAndScopes(Request request, ServiceMethodInfo servic
...
     /**
+     * Extract the EndUsername from accessToken.
+     *
+     * @param accessToken the access token
+     * @return loggedInUser if the token is a valid token
+     */
+    private String getEndUserName(String accessToken) throws APIMgtSecurityException {
+        String loggedInUser;
+        loggedInUser = validateToken(accessToken).getEndUserName();
+        return loggedInUser.substring(0, loggedInUser.indexOf("@"));
```
Comment (bhathiya) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api.commons/src/main/java/org/wso2/carbon/apimgt/rest/api/common/impl/OAuth2Authenticator.java]: should this be the last @ of `loggedInUser` to work with usernames such as user@gmail.com@carbon.super?

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.commons/src/main/java/org/wso2/carbon/apimgt/rest/api/common/util/RestApiUtil.java [hunk 2]
```diff
@@ -61,12 +63,11 @@
...
      * @return The current logged in user.
+     * @param request
      */
-    public static String getLoggedInUsername() {
-//        CarbonContext carbonContext = CarbonContext.getCurrentContext();
-//        Principal principal = carbonContext.getUserPrincipal();
-//        return principal.getName();
-        return "admin";
+    public static String getLoggedInUsername(Request request) {
+        String loggedInUser = request.getProperty(LOGGED_IN_USER).toString();
```
Comment (bhathiya) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api.commons/src/main/java/org/wso2/carbon/apimgt/rest/api/common/util/RestApiUtil.java]: better to have a null check.

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.commons/src/main/java/org/wso2/carbon/apimgt/rest/api/common/impl/OAuth2Authenticator.java (hunk 1 above)
Comment (Krishanx92) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api.commons/src/main/java/org/wso2/carbon/apimgt/rest/api/common/impl/OAuth2Authenticator.java]: I think IS not give permission to create usernames with special characters such as @.but anyway I'll change that accordingly to last @.

Comment (bhathiya) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api.commons/src/main/java/org/wso2/carbon/apimgt/rest/api/common/impl/OAuth2Authenticator.java]: There can be email usernames.

**Pull request merged**